pep-0497-externals.txt

```diff
@@ -0,0 +1,359 @@
+        install_requires=['yaml', 'scipy[atlas]', 'lib!libblas1.so'],
+        setup_requires=[
+            'pyfortran',
+            'bin!fortran:dist_name=="ubuntu"',
+            'include!kernel.h',
+            'include!liblas.h',
+          ],
+    )
+
+
+If the user does not have the required dependencies, the user will be presented
+with a useful install message and a process that exits early, as opposed to
+relying on the compiler to fail and provide an error message that may not be
+readily understood by most users.
+
+The mechanism of referencing failed dependencies such as 'kernel.h' to useful
+user messages regarding the actual operating system packages that are relevant
+is discussed in the implementation section, along with the rationale for
+handling that referencing outside of setup.py.
+
+Rationale
+=========
+
+Differences between environments have implications for the building, testing,
+installing and execution of software. There are two sides to this coin. First,
+in some environments, additional Python packages are needed. Secondly, in some
+environments, additional non-Python packages need to be present, such as header
+files, libraries, scripts or other binaries. The syntax proposed here allows
+the description of both sorts of requirements in a single syntax.
+
+Setuptools and pip should be modified to support the use case of installing
+the appropriate python packages as required depending on the specification.
+
+They should also support aborting an install with a useful error message if
+required dependecies are not installed in the environment. The responsibility
+for actually installing those required packages lies with the user.
+
+This would allow automatic installation of environment-specific Python
+dependencies and pave the way to improving how users go about installing
+non-Python prerequisites.
+
+Ideally, and error message will display the package manager command required
+to add the missing environment dependencies.
+
+It is expected that this feature would be commonly used, particularly in
+scientific applications which frequently have complex interactions with
+externally provided libraries. An example would be needing versions of
+numpy and scipy, both of which must have been compiled to be aware of the ATLAS
+compiled set of linear algebra libraries for performance reasons. This sounds
+esoteric but is, in fact, a routinely encountered situation which drives
+people towards using the alternative packaging for scientific python
+environments.
+
+Syntax
+======
+
+The valid context specifiers are taken from the environment markers library.
+This command says that when the operating system matches 'ubuntu', then add
+the yaml==0.3 requirement to the list.
+
+Meta: I am not a syntax wizard and will need help from others to properly specify
+the syntax as a formal grammar if that is required.
+
+Essentially, the "!" separator is used to flag key-value pairs which will allow
```

Comment:
"The proposed syntax is an external object type, a "!", and then an external object name. The object type can be one of:

- include
- bin
- lib

The object name should be the name of the required thing - search paths are presumed to be used to find the thing.